flips/20211007-transaction-fees.md

```diff
@@ -0,0 +1,264 @@
+
+
+### Execution effort
+
+The execution effort represents the effort needed to execute the transaction script and to handle the execution results. Because the execution path of the script depends on the current execution state, it is impossible to definitively predict the execution effort before the actual on chain execution of the transaction.
+
+The way execution effort is defined, means that the lower bound for execution effort is 0. Any constant part would just get counted under the inclusion effort.
+
+The upper bound of the execution fees (and thus also for transaction fees; assuming the inclusion fees are known) is indirectly set in the transaction by setting the maximum execution effort limit (<img src="https://render.githubusercontent.com/render/math?math=E_c%5E%5Ctext%7Bmax%7D">) (in the current implementation this is known as the gas limit). 
+
+If the transaction execution effort limit is reached during the execution of the transaction the execution stops and the state changes made up to that point are dropped. The transaction fees for this transaction are still collected at the maximum execution effort limit.
+
+Currently the execution effort is calculated by adding 1 for every loop or function call. This could be sufficient for the first implementation. In the future the calculation of the execution effort would be changed so that function calls would have a different execution effort cost assigned to them according to the time complexity of the method.
+
+### Failing transactions
+
+There are a few reasons why a transaction fails, which can be split into four categories, according to who finally gets charged for the transaction fees and how much do they get charged.
+
+1. Transaction failed because the payers signature was incorrect, or the payer does not have enough funds to cover maximum transaction costs (transaction costs with the execution effort set to the execution effort limit).
+2. Transaction failed before the transaction script execution started (some non-payer signature was incorrect, or the sequence number was incorrect)
+3. Transaction failed during transaction script execution, during fee deduction or during the storage used check. 
+4. Transaction failed because the execution effort limit was reached.
+
+As discussed below, in scenario 1. the transaction shouldn't have passed the access node in the first place. However since it has and there is no way to charge the payer, the access node account is charged with the transaction fees. The transaction fees are computed with the execution effort set to 0.
+
```

Comment:
### Follow-up summarizing discussion with Dete

It is a very subtle combination of: 
1. A variant of the verifier's dilemma on the execution nodes, where they stop checking for transaction duplicates because the collectors never produce duplicates. In this mind set, every conflicting sequence number is the fault of the payer. Checking for duplicates is a resource-intensive task. There is no mechanism to preventing execution nodes becoming lazy and falling for the Verifier's Dilemma. 
2. However, as soon as Execution Nodes stop checking, a byzantine collector cluster could double-charge any client account by duplicating transactions (forcing sequence number collisions without the client being at fault). 

The threshold of this attack is relatively high, as it requires all execution nodes to be lazy (though not intentionally malicious) _plus_ a byzantine collector cluster. Side note: Reaching 1/3 weight is sufficient to pull off a _single_ replication attack of every submitted transaction. To fully bleed an arbitrary account, a malicious supermajority in a cluster would be required.